### PR TITLE
Add progress bar components

### PR DESCRIPTION
**Pull request by 3ace**
This PR add a linear progress bar and a circular progress bar component. 

Preview:

preview

### CODE REVIEW

File: progress_bar_circular.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+)
+
+// CircularProgressBar is a component that will render circular progress bar component.
+type CircularProgressBar struct {
+	BackgroundStyle render.Style
+	ForegroundStyle render.Style
+	LabelStyle      render.Style
+	ColorPalette    render.ColorPalette
+
+	// If the value is true then the progress bar would rendered counter clockwise.
```
Comment (sampila) [inline progress_bar_circular.go:19]: Probably should be
// Reversed if the value is `true` then the progress bar would rendered counter clockwise.

File: progress_bar_circular.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	Reversed bool
+
+	// Size of the progress bar in width and height.
```
Comment (sampila) [inline progress_bar_circular.go:22]: // size ....

File: progress_bar_circular.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	size int
+	dpi  float64
+
+	// Progress values which should be between 0.0 - 1.0.
+	progress float64
```
Comment (sampila) [inline progress_bar_circular.go:27]: // progress ...

File: progress_bar_circular.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+	// Label to be displayed in the center of the chart.
```
Comment (sampila) [inline progress_bar_circular.go:29]: // label ....

File: progress_bar_linear.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	"io"
+	"math"
+
+	"github.com/unidoc/unichart/mathutil"
+	"github.com/unidoc/unichart/render"
+)
+
+// LinearProgressBar is a component that will render progress bar component.
+type LinearProgressBar struct {
+	// Style for the background bar.
```
Comment (sampila) [inline progress_bar_linear.go:13]: // BackgroundStyle ....

File: progress_bar_linear.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+	BackgroundStyle render.Style
+
+	// Style for the foreground bar.
```
Comment (sampila) [inline progress_bar_linear.go:16]: // ForegroundStyle ...

File: progress_bar_linear.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+	ForegroundStyle render.Style
+
+	//
```
Comment (sampila) [inline progress_bar_linear.go:19]: Empty comment?

File: progress_bar_linear.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+	ColorPalette render.ColorPalette
+
+	// Flag to enable rounded edge at the start of the bar.
```
Comment (sampila) [inline progress_bar_linear.go:22]: // RoundedEdgeStart ...

File: progress_bar_linear.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+	RoundedEdgeStart bool
+
+	// Flag to enable rounded edge at the end of the bar.
```
Comment (sampila) [inline progress_bar_linear.go:25]: // RoundedEdgeEnd ...

File: progress_bar_linear.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+	RoundedEdgeEnd bool
+
+	height int
+	width  int
+	dpi    float64
+
+	// Progress values which should be between 0.0 - 1.0.
```
Comment (sampila) [inline progress_bar_linear.go:32]: // progress ...

File: progress_bar_circular.go [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+func (cp *CircularProgressBar) SetWidth(width int) {
+}
+
+// Height returns the chart height.
+func (cp *CircularProgressBar) Height() int {
+	return cp.Size()
+}
+
+// SetHeight sets the chart height.
+func (cp *CircularProgressBar) SetHeight(height int) {
```
Comment (sampila) [inline progress_bar_circular.go:94]: Empty function?

File: progress_bar_circular.go [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+	return cp.label
+}
+
+// Width returns the chart width.
+func (cp *CircularProgressBar) Width() int {
+	return cp.Size()
+}
+
+// SetWidth sets the chart width.
+func (cp *CircularProgressBar) SetWidth(width int) {
```
Comment (sampila) [inline progress_bar_circular.go:85]: Empty function? or this is needed to implements interface?

**Review by sampila — CHANGES_REQUESTED** [review]
Great works on this, added minor comments.

File: progress_bar_circular.go (hunk 12 above)
Comment (3ace) [reply progress_bar_circular.go:85]: yes, these empty functions is needed to fulfill the interface requirements. I'll add more info in the comments later.

**Comment by 3ace** [comment]
@sampila I've update the comments section as you suggested. Thanks

**Comment by 3ace** [comment]
Added a label support in the linear progress bar.

File: progress_bar_circular.go [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+	}
+	return cp.dpi
+}
+
+// SetDPI sets the DPI for the progrss bar.
+func (cp *CircularProgressBar) SetDPI(dpi float64) {
+	cp.dpi = dpi
+}
+
+// Width returns the chart size or the default value.
```
Comment (sampila) [inline progress_bar_circular.go:65]: // Size ...

File: progress_bar_circular.go [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+func (cp *CircularProgressBar) Size() int {
+	if cp.size == 0 {
+		return defaultChartWidth
+	}
+	return cp.size
+}
+
+// SetWidth sets the chart size.
```
Comment (sampila) [inline progress_bar_circular.go:73]: // SetSize ...

File: progress_bar_circular.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+func (cp *CircularProgressBar) SetSize(size int) {
+	cp.size = size
+}
+
+func (cp *CircularProgressBar) SetLabel(label string) {
```
Comment (sampila) [inline progress_bar_circular.go:78]: Need to add comment regarding this function

File: progress_bar_circular.go [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+	cp.label = label
+}
+
+func (cp *CircularProgressBar) GetLabel() string {
```
Comment (sampila) [inline progress_bar_circular.go:82]: Need to add function regarding this function

File: progress_bar_linear.go [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+	}
+	return lp.height + lp.barYPos + lp.bottomInfoHeight
+}
+
+// SetHeight sets the chart height.
+func (lp *LinearProgressBar) SetHeight(height int) {
+	lp.height = height
+}
+
+func (lp *LinearProgressBar) SetLabel(label string) {
```
Comment (sampila) [inline progress_bar_linear.go:106]: Need to add comment regarding this exported function

File: progress_bar_linear.go [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+	lp.label = label
+}
+
+func (lp *LinearProgressBar) GetLabel() string {
```
Comment (sampila) [inline progress_bar_linear.go:110]: Need to add comment regarding this exported function

**Review by sampila — COMMENTED** [review]
Added minor comments

**Comment by 3ace** [comment]
@sampila comments updated in the latest commit.

**Review by sampila — APPROVED** [review]
Looks good to me

File: progress_bar_linear.go [hunk 19]
```diff
@@ -0,0 +1,283 @@
...
+	x := float64(lp.Width()) * lp.progress
+
+	if lp.CustomBottomInfo != nil {
+		lp.bottomInfoHeight = lp.CustomBottomInfo(r, int(x))
+
+		return
+	}
+}
+
+// Render renders the progrss bar with the given renderer to the given io.Writer.
```
Comment (gunnsth) [inline progress_bar_linear.go:266]: typo in "progrss"

File: progress_bar_circular.go [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+
+	if labelStyle.Hidden || cp.label == "" {
+		return
+	}
+
+	fgStrokeWidth := int(labelStyle.StrokeWidth)
+	render.Text.DrawWithin(r, cp.label, render.NewBox(fgStrokeWidth, fgStrokeWidth, cp.Size()-fgStrokeWidth, cp.Size()-fgStrokeWidth), labelStyle)
+}
+
+// Render renders the progrss bar with the given renderer to the given io.Writer.
```
Comment (gunnsth) [inline progress_bar_circular.go:202]: typo in "progrss"

**Comment by 3ace** [comment]
@gunnsth I've fixed those typos and update the readme file.

**Review by gunnsth — COMMENTED** [review]
@3ace For the circular progress bar, is it possible to display the percentage like "95%" or "95.9%" inside the circle instead of the A?  Is that somehow customizable?

**Comment by 3ace** [comment]
@gunnsth that text (and it's size or color) is customizable via `label` parameter. User could also hide that label if they want. The default is just empty text

**Review by gunnsth — APPROVED** [review]
LGTM

**Pull request merged**